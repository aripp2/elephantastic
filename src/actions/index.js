export const setRandom = pup =>({
  type: 'SET_RANDOM',
  pup
})

export const throwError = error => ({
  type: 'THROW_ERROR',
  error
})

export const setFavs = favs => ({
  type: 'SET_FAVS',
  favs
})