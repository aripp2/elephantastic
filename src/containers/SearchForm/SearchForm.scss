@import './src/variables.scss';

.SearchContainer {
  height: 100vh;
  width: 100%;
  font-family: $body-font;
  display: flex;
}

aside {
  height: 100vh;
  width: 25%;
  border-right: 2px solid black;
}

select {
  height: 40px;
  width: 85%;
  font-family: $body-font;
  font-size: 1em;
}

h3 {
  font-size: 1.5em;
}

h4 {
  font-size: 1.7em;
}

span {
  font-weight: 700;
}

h5 {
  font-size: 1.2em;
  margin: 20px;
}
