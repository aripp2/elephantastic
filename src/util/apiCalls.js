const xKey = 'fdae1377-8b39-4c2b-a87f-38d9ef268562';
const getOptions = {
    method: 'GET',
    headers: {
      'x-api-key': xKey
    }
  };

export const getRandom = async() => {
  const url = 'https://api.thedogapi.com/v1/images/search';
  const response = await fetch(url, getOptions)
  if(!response.ok) {
    throw Error('There was an issue getting a pup, try again later.')
  }
  const random = await response.json();
  return random[0]

}

export const postVote = async (imgId, value) => {
  const url = 'https://api.thedogapi.com/v1/votes';
  const options = {
    method: 'POST',
    body: JSON.stringify({
      image_id: imgId,
      value: value
    }),
    headers: {
      'Content-type': 'application/json',
      'x-api-key': xKey
    }
  }

  const response = await fetch(url, options);
  if(!response.ok) {
    throw Error('Unable to submit your vote.')  
  }
  const votes = await response.json()
  console.log('postVote in apiCalls', votes)
}

export const getBreeds = async () => {
  const url = 'https://api.thedogapi.com/v1/breeds';
  const response = await fetch(url, getOptions);
  if(!response.ok) {
    console.log(response)
  }
  const breeds = await response.json();
  // console.log('getBreeds in apiCalls', breeds)
  return breeds;
}

export const getFavorites = async() => {
  const url = 'https://api.thedogapi.com/v1/favourites';
  const response = await fetch(url, getOptions)
  if(!response.ok) {
    console.log(response)
  }
  const favs = await response.json();
  console.log('favs', favs)
  return favs;
}

export const addFavorite = async(imgId) => {
  const url = 'https://api.thedogapi.com/v1/favourites';
  const options = {
    method: 'POST',
    body: JSON.stringify({
      image_id: imgId
    }),
    headers: {
      'Content-type': 'application/json',
      'x-api-key': xKey
    }
  } 

  const response = await fetch(url, options)
  if(!response.ok) {
    throw Error('Unable to add favorite.')
  }
}

export const deleteFavorite = async(favId) => {
  const url = `https://api.thedogapi.com/v1/favourites/${favId}`
  const options = {
    method: 'DELETE',
    headers: {
      'Content-type': 'application/json',
      'x-api-key': xKey
    }
  }
  const response = await fetch(url, options)
  if(!response.ok) {
    throw Error('Unable to remove this favorite.')
  }
}

export const getBreedImages = async(breedId) => {
  const url = `https://api.TheDogAPI.com/images/search?breed_id=${breedId}`
  const response = await fetch(url, getOptions)
  if(!response.ok) {
    console.log('selected breed response', response)
  }
  const selected = await response.json()
  console.log('selected in api', selected)
  return selected
}





