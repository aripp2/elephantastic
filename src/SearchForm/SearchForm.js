import React, { Component } from 'react';
import { connect } from 'react-redux';
import { getBreedImages } from '../util/apiCalls';
import { throwError, updateLoading } from '../actions';
import SearchContainer from '../SearchContainer/SearchContainer';
import './SearchForm.scss';

export class SearchForm extends Component {
  constructor() {
    super();
    this.state = {
      value: '',
      selectedBreed: null,
      breedImages: []
    }
  }

  changeHandler = (e) => {
    const { breeds } = this.props
    const selected = breeds.find(breed => breed.name === e.target.value)
    this.setState({ selectedBreed: selected })
    this.getSelectedImages(selected.id)
  }

  getSelectedImages = async(id) => {
    const { throwError, updateLoading } = this.props;
    updateLoading(true);
    try {
      const breedImages = await getBreedImages(id)
      console.log(breedImages)
      this.setState({ breedImages })
      updateLoading(false)
    } catch ({ message }){
      updateLoading(false)
      throwError(message)
    }
  }

  render() {
    const { selectedBreed, breedImages } = this.state;
    const { errorMsg, isLoading } = this.props;
    const { breeds } = this.props
    const breedList = breeds.map(breed => {
      return <option 
        key={breed.id} 
        value={breed.name}
        >{breed.name}</option>
    })

    return (
      <div>
        <aside>
          <form>
            <h3>Choose a Breed:</h3>
            <select
              value={selectedBreed}
              onChange={this.changeHandler}
            >
              { breedList }
            </select>
          </form>
          {selectedBreed && 
            <article>
              <h4>{selectedBreed.name}</h4>
              <p>Breed Group: {selectedBreed.breed_group}</p>
              <p>Bred For: {selectedBreed.bred_for}</p>
              <p>Average Height: {selectedBreed.height.imperial} in</p>
              <p>Average Weight: {selectedBreed.weight.imperial} lbs</p>
              <p>Average Life Span: {selectedBreed.life_span}</p>
              <p>{selectedBreed.temperament}</p>
          </article>}
        </aside>
        {errorMsg && <h2>{errorMsg}</h2>}
        {isLoading && <img src='https://cdn.dribbble.com/users/5976/screenshots/3930514/happy_dog_puppy_tongue_logo_design_symbol_by_alex_tass.gif' alt='loading gif'/>}
        <SearchContainer breedImages={breedImages}/>
      </div>
    )
  }
}

export const mapStateToProps = ({ breeds, favorites, errorMsg, isLoading }) => ({
  breeds, 
  favorites,
  errorMsg,
  isLoading
});

export const mapDispatchToProps = dispatch => ({
  throwError: error => dispatch(throwError(error)),
  updateLoading: bool => dispatch(updateLoading(bool))
})

export default connect(mapStateToProps, mapDispatchToProps)(SearchForm);