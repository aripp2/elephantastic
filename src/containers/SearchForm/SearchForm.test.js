import React from 'react';
import { shallow } from 'enzyme';
import { SearchForm, mapStateToProps, mapDispatchToProps } from './SearchForm';
import { throwError, updateLoading } from '../../actions';
import { getBreedImages } from '../../util/apiCalls';

jest.mock('../../util/apiCalls.js');

const mockBreeds = [
  {
    bred_for: "Sled pulling",
    breed_group: "Mixed",
    height: {imperial: "23 - 26", metric: "58 - 66"},
    id: 8,
    life_span: "10 - 13 years",
    name: "Alaskan Husky",
    temperament: "Friendly, Energetic, Loyal, Gentle, Confident",
    weight: {imperial: "38 - 50", metric: "17 - 23"}
  },
  {
    bred_for: "Hauling heavy freight, Sled pulling",
    breed_group: "Working",
    height: {imperial: "23 - 25", metric: "58 - 64"},
    id: 9,
    life_span: "12 - 15 years",
    name: "Alaskan Malamute",
    temperament: "Friendly, Affectionate, Devoted, Loyal, Dignified, Playful",
    weight: {imperial: "65 - 100", metric: "29 - 45"}
  }];
  const mockBreedImages = [
    {
      breeds: [{}],
      height: 2592,
      id: "GhtSdrW29",
      url: "https://cdn2.thedogapi.com/images/GhtSdrW29.jpg",
      width: 3888
    },
    {
      breeds: [{}],
      height: 668,
      id: "zEQtlI8h1",
      url: "https://cdn2.thedogapi.com/images/zEQtlI8h1.jpg",
      width: 960
    }
  ];
  const mockError = '';
  getBreedImages.mockImplementation(() => Promise.resolve(mockBreedImages));

describe('SearchForm', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<SearchForm 
      breeds={mockBreeds}
      errorMsg={mockError}
      isLoading={false}
      throwError={jest.fn()}
      updateLoading={jest.fn()}
      />)
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should update state when change happens', () => {
    const mockEvent = { target: { value: 'Alaskan Malamute'}}
    wrapper.instance().changeHandler(mockEvent) 
    expect(wrapper.state('selectedBreed')).toEqual(mockBreeds[1]);
  });

  it('should call getSelectedImages when changeHandler is called', () => {
    wrapper.instance().getSelectedImages = jest.fn();
    const mockEvent = { target: { value: 'Alaskan Malamute'}}
    wrapper.instance().changeHandler(mockEvent);
    expect(wrapper.instance().getSelectedImages).toHaveBeenCalledWith(9)
  });

  it('should call getBreedImages when getSelectedImages is called', () => {
    wrapper.instance().getSelectedImages(9);
    expect(getBreedImages).toHaveBeenCalled();
  });

  // it('should catch an error if response fails', () => {
  //   getBreedImages.mockImplementation(() => Promise.reject({message: 'Fail'}))
  //   expect(throwError()).toHaveBeenCalledWith('Fail')
  // })

});

describe('mapStateToProps', () => {

  it('should return an object of errorMsg and isLoading', () => {
    const mockState = {
      favorites: [],
      randomPup: {},
      errorMsg: '',
      isLoading: false,
      breeds: mockBreeds
    };
    const expected = {
      errorMsg: '',
      isLoading: false,
      breeds: mockBreeds
    }
    const mappedProps = mapStateToProps(mockState);
    expect(mappedProps).toEqual(expected);
  });
});

describe('mapDispatchToProps', () => {

  const mockDispatch = jest.fn();

  it('calls dispatch with a throwError action', () => {
    const mockError = 'An Error';
    const actionToDispatch = throwError(mockError);
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.throwError(mockError);
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });

  it('calls dispatch with a updateLoading action', () => {
    const mockLoading = true;
    const actionToDispatch = updateLoading(mockLoading);
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.updateLoading(mockLoading);
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });

});

