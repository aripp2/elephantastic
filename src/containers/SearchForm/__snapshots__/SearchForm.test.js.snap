// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm should match the snapshot 1`] = `
<div
  className="SearchContainer"
>
  <aside>
    <form>
      <h3>
        Choose a Breed:
      </h3>
      <select
        onChange={[Function]}
        value=""
      >
        <option
          key="8"
          value="Alaskan Husky"
        >
          Alaskan Husky
        </option>
        <option
          key="9"
          value="Alaskan Malamute"
        >
          Alaskan Malamute
        </option>
      </select>
    </form>
  </aside>
  <SearchContainer
    breedImages={Array []}
  />
</div>
`;
